posts/Swift-Optionals-Are-Monads-Too.md

```diff
@@ -0,0 +1,264 @@
+
+A simple example:
+
+```swift
+var string: String? = "Yay! I contain a value."
+var emptyString: String?
+
+// We unwrap our Optional before we print it out
+if let string = string {
+  print(string)
+}
+
+// Try to unwrap our emptyString variable
+if let emptyString = emptyString {
+  // we won't make it here
+  print(emptyString)
+} else {
+  // here we can handle the case where the Optional
+  // doesn't contain a value
+  print("Your string was empty")
+}
+```
+---
+
+# What the Functor
+
+So what is a functor? It sounds complicated and only adding to it's complicated nature, it's roots are in Mathematics (Category Theory to be a bit more specific). But let's try and define it in terms a programmer understands instead of mathematical terms.
+
+> A functor is any type that defines how map applies to it. 
+
+It just so happens that Optional defines this very method. Let's look again at the declaration for Optional, but cut down on the noise a bit.
+
+```swift
+public enum Optional<T> {
+    // ...
+
+    public func map<U>(f: (T) -> U) -> U?
+
+    // ...
+}
+```
+
+If we break down Optional's implementation of map, it's a function that accepts another function as it's only argument. This other function, we'll call it f, accepts an object of the type T and returns an object of the type U. And finally, map returns U? which is an Optional of the type U.
+
+A more concrete example of map being used looks something like this:
+
+```swift
+let value: Int? = 42
+let anotherValue = value.map { "\($0 * 2)" } // anotherValue is initialized to "84"
+```
+
+In this case, if we fill in the generic blanks, T is Int and U is String. f accepts an Int and returns a String. And finally, map returns a String?. In our case it's value is "84".
+
+The more interesting example of map in use is when value is nil.
+
+```swift
+let value: Int?
+let anotherValue = value.map { "\($) * 2)" } // anotherValue is initialized to nil
+```
+
+When our Optional doesn't contain a value, we skip over calling f with our missing value and just return .None. So if I were guessing how Apple implemented map on the Optional class, I'd figure it would look similar to this:
+
+```swift
+func map<U>(@noescape f: (Wrapped) throws -> U) rethrows -> U? {
+  switch self {
+  case .Some(let x): return try f(x)
+  case .None: return .None
+  }
+}
+```
+
+So Swift's Optional is a functor because it meets the requirement of defining the map method.
+
+--- 
+
+# Applicative Functor
+
+Armed with the knowledge of what a Functor is, next up is Applicative Functor. Let's start with a definition in programmer terms.
+
+> An applicative functor is any type that defines apply.
+
+Looking back at the declaration for Optional, we don't find any methods defining apply. Well hmm, by default Optionals aren't applicative functors. But wait! We can add an extension to this built in type and make it work. 
```

Comment:
It's totally acceptable but you don't need to use words like 'Well hmm,' you can just do 'Hmm'.  Or 'But wait!', could be 'Wait!'.  Again this is just a style thing.